src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -1324,45 +1335,42 @@ bool LibraryCallKit::inline_vector_mem_masked_operation(bool is_store) {
     set_sp(old_sp);
     return false;
   }
+  if (using_byte_array) {
+    // Reinterpret the vector mask to byte type.
+    const TypeVect* from_mask_type = TypeVect::makemask(elem_bt, num_elem);
+    const TypeVect* to_mask_type = TypeVect::makemask(mem_elem_bt, mem_num_elem);
+    mask = gvn().transform(new VectorReinterpretNode(mask, from_mask_type, to_mask_type));
+  }
+
+  const TypeVect* vt = TypeVect::make(elem_bt, num_elem);
+  const TypeVect* mem_vt = TypeVect::make(mem_elem_bt, mem_num_elem);
 
   if (is_store) {
-    Node* val = unbox_vector(argument(7), vbox_type, elem_bt, num_elem);
-    if (val == NULL) {
-      if (C->print_intrinsics()) {
-        tty->print_cr("  ** unbox failed vector=%s",
-                      NodeClassNames[argument(7)->Opcode()]);
-      }
-      set_map(old_map);
-      set_sp(old_sp);
-      return false; // operand unboxing failed
-    }
     set_all_memory(reset_memory());
 
     if (using_byte_array) {
       // Reinterpret the incoming vector to byte vector.
-      const TypeVect* to_vect_type = TypeVect::make(mem_elem_bt, mem_num_elem);
-      val = gvn().transform(new VectorReinterpretNode(val, val->bottom_type()->is_vect(), to_vect_type));
-      // Reinterpret the vector mask to byte type.
-      const TypeVect* from_mask_type = TypeVect::makemask(elem_bt, num_elem);
-      const TypeVect* to_mask_type = TypeVect::makemask(mem_elem_bt, mem_num_elem);
-      mask = gvn().transform(new VectorReinterpretNode(mask, from_mask_type, to_mask_type));
+      val = gvn().transform(new VectorReinterpretNode(val, val->bottom_type()->is_vect(), mem_vt));
+    }
+
+    Node* vstore = NULL;
+    if (supports_predicate) {
+      // Generate masked store vector node if predicate feature is supported.
+      vstore = gvn().transform(new StoreVectorMaskedNode(control(), memory(addr), addr, val, addr_type, mask));
+    } else {
+      // Use the vector blend to implement the masked store. The biased elements are the original
+      // values in the memory.
+      Node* mem_val = gvn().transform(LoadVectorNode::make(0, control(), memory(addr), addr, addr_type, mem_num_elem, mem_elem_bt));
```

Comment:
I'm sorry to say it, but I am pretty sure this is an invalid optimization.
See top-level comment for more details.

Reply:
Thanks for your comments! Yeah, this actually influences something due to the Java Memory Model rules which I missed to consider more. I will try the scatter ways instead. Thanks so much!

Reply:
Yes, phantom store can write back stale unintended value any may create problem in multithreded applications since blending is done with an older loaded value.